index.js

```diff
@@ -16,45 +16,92 @@ var nightmare = Nightmare(),
     maxQueue = Infinity,
     queue = new Queue(maxConcurrent, maxQueue);
 
-// Use bodyParser middleware to
-// parse json responses.
-app.use(bodyParser.json());
-
-app.post('/render', function(req, res) {
-  queue.add(function() {
-    return render(req.body)
-  }).then(function(output) {
-    res.sendFile(output);
-  });
-});
+// Create the rendering engine for a new render call
+function render(data) {
+    var id = uuid.v4(),
+        output = '/tmp/' + id + '.pdf',
+        router = express.Router(),
+        deferred;
 
-app.listen(config.port);
+    app.use('/' + id, router);
 
-function render(data) {
-  var id = uuid.v4(),
-      output = '/tmp/' + id + '.pdf',
-      router = express.Router(),
-      deferred;
-
-  app.use('/' + id, router);
-
-  router.use(express.static(config.document_path));
-  router.get('/data.json', function(req, res) { res.json(data); });
-  router.get('/data', function(req, res) { res.type("text/javascript").send("var data = " + JSON.stringify(data)); });
-  router.get('/defer', function(req, res) {
-    deferred = new Promise(function(resolve, reject) {
-      router.get('/done', function(req, res) {
-        resolve();
+    router.use(express.static(config.document_path));
+    router.get('/data.json', function (req, res) {
+        log('debug', 'Serving data.json for id ' + id);
+        res.json(data);
+    });
+    router.get('/data', function (req, res) {
+        log('debug', 'Serving data in javascript for id ' + id);
+        res.type("text/javascript").send("var data = " + JSON.stringify(data));
+    });
+    router.get('/defer', function (req, res) {
+        deferred = new Promise(function (resolve, reject) {
+            router.get('/done', function (req, res) {
+                log('debug', 'Document called done. id: ' + id);
+                resolve();
+                res.status(200).end();
+            });
+        });
         res.status(200).end();
-      });
     });
 
-    res.status(200).end();
-  });
+    return Promise
+        .resolve(nightmare.goto('http://localhost:' + getMyPort() + '/' + id))
+        .then(function () {
+            return deferred;
+        })
+        .then(function () {
+            // Converting the HTML to PDF with nightmare
+            log('debug', 'Outputting the html to pdf');
+            return nightmare.pdf(output, {marginsType: 2});
+        })
+        .then(function () {
+            // Return the location of the document
+            return output;
+        });
+}
+
+// Log some stuff to the console
+function log(level, message) {
+    console.log("[" + level.toUpperCase() + "]: " + message);
+}
+
+// Ge the servers configured port (either config file or environment variable PORT)
+function getMyPort(withLog) {
```

Comment:
so I can reuse it (for the server startup and the location where we send nightmare to the server)